Clarify server CONNECTION_CLOSE with Handshake (#2688)
quicwg/base-drafts: draft-ietf-quic-transport.md

protocol can use an CONNECTION_CLOSE frame with an appropriate error code to
signal closure.

When sending CONNECTION_CLOSE, the goal is to ensure that a peer will process
the frame.  Generally, this means sending the frame in a packet with the highest
level of packet protection to avoid the packet being discarded.  However, during
the handshake, it is possible that more advanced packet protection keys are not
available to a peer, so the frame MAY be replicated in a packet that uses a
lower packet protection level.

If the handshake is confirmed, endpoints MUST send any CONNECTION_CLOSE frames
in a 1-RTT packet.  Prior to handshake confirmation, a peer might not have
1-RTT keys, so endpoints SHOULD send CONNECTION_CLOSE frames in a Handshake
packet.  If the endpoint does not have Handshake keys, it SHOULD send
CONNECTION_CLOSE frames in an Initial packet.

> The server may not know whether the client has Handshake keys.  In order to

Reviewer: ```suggestion
A client will always know whether the server has Handshake keys (see {{discard-initial}}), but it is possible that a server does not know whether the client has Handshake keys.
```

protocol can use an CONNECTION_CLOSE frame with an appropriate error code to
signal closure.

> When sending CONNECTION_CLOSE, the goal is to ensure that a peer will process

Reviewer: ```suggestion
When sending CONNECTION_CLOSE, the goal is to ensure that the peer will process
```

the frame.  Generally, this means sending the frame in a packet with the highest
level of packet protection to avoid the packet being discarded.  However, during
the handshake, it is possible that more advanced packet protection keys are not

> available to a peer, so the frame MAY be replicated in a packet that uses a

Reviewer: ```suggestion
available to the peer, so the frame MAY be replicated in a packet that uses a
```

> lower packet protection level.

Reviewer 2: This paragraph doesn't really say anything that the next paragraph doesn't cover. Also, `a peer` sounds like an endpoint can have multiple peers.

Reviewer: I missed this earlier, but I agree. I'm fine with dropping this paragraph entirely. It seems like it's simply repeating what's in the next para, but in more general terms.

Author: I don't have a strong opinion, but I see your point, this was @martinthomson addition, so I'll let him comment.

Reviewer 3: The point was to articulate the principles behind the design.  It might be that the mechanics are sufficiently precise that that is not necessary, but we did have #2151.  @ianswett's changes didn't really directly address that.

Reviewer 3: If people want to fix this, please open a followup issue.

> If the handshake is confirmed, endpoints MUST send any CONNECTION_CLOSE frames

Reviewer 2: ```suggestion
When the handshake is confirmed, an endpoint MUST send any CONNECTION_CLOSE frames
```

Author: That makes it sound a bit too much like one should close the connection upon completing the handshake.  I think If works better, though "After" could also work.

> in a 1-RTT packet.  Prior to handshake confirmation, a peer might not have

Reviewer 2: ```suggestion
in a 1-RTT packet.  Prior to handshake confirmation, the peer might not have
```

> 1-RTT keys, so endpoints SHOULD send CONNECTION_CLOSE frames in a Handshake

Reviewer 2: ```suggestion
1-RTT keys, so the endpoint SHOULD send CONNECTION_CLOSE frames in a Handshake
```

If the connection has been successfully established, endpoints MUST send any
CONNECTION_CLOSE frames in a 1-RTT packet.  Prior to connection establishment a
peer might not have 1-RTT keys, so endpoints SHOULD send CONNECTION_CLOSE frames

> in a Handshake packet.If the endpoint does not have Handshake keys, it MAY

Reviewer 4: ```suggestion
in a Handshake packet.  If the endpoint does not have Handshake keys, it MAY
```

> send CONNECTION_CLOSE frames in an Initial packet.

Reviewer: This needs to be only a statement of fact that describes the exception to the preceding SHOULD. "If the endpoint does not have Handshake keys, it sends CONNECTION_CLOSE frames in an Initial packet"

Author: I'm confused why we have a MUST and a SHOULD for the other two cases, but you're advising no normative text for the Initial case?  Am I missing something?

> The server can have Handshake keys but not know that the client has Handshake

Reviewer 4: The `but not know` reads a little weird to me for some reason. Perhaps something along the lines of `but not be certain the client does.`?

Author: Rewritten

> The server may not know whether the client has Handshake keys.  In order to

Reviewer: ```suggestion
It is possible that an endpoint does not know whether the peer has Handshake keys.  In order to
```

Author: I'm fairly sure server to client is the only case two connection closes may need to be sent, which is why it's written with server an client.

Author: I have in mind 17.2.2.1 in Transport: https://tools.ietf.org/html/draft-ietf-quic-transport-20#section-17.2.2.1

"A client stops both sending and processing Initial packets when it sends its first Handshake packet."

This clearly indicates that a client will almost never be processing Initial and Handshake packets at once, so if you only send one connection close, some clients will drop it.  In all other cases with #2673 landed, the client and server should both still have Handshake keys even if they also have 1-RTT keys?

Is there something about TLS I'm implicitly assuming as well?  I can imagine I am, but I don't know what it is.

Reviewer 3: Ahh, the two docs blur together in my mind.  This is OK.  I would suggest that you need to address both transitions here though.  The transition to Handshake and the transition to 1-RTT have the same sort of uncertainty.

Author: The handshake to 1-RTT transition has an overlap where the peer has both sets of keys, unlike Initial to Handshake, so I think using handshake confirmed as the signal here provides what we want?

> keys.  In order to guarantee a CONNECTION_CLOSE is processed, it has to send

Reviewer 4: Should `has`  be `SHOULD`? If you're going to send a CONNECTION_CLOSE then you want it to be received. If the server isn't certain the client has the handshake keys, it really should send both, IMO.

Author: I tend to agree, but I was attempting to write non-normative text.  I'll change that and see how it reads.

> guarantee a CONNECTION_CLOSE is processed, it SHOULD send a CONNECTION_CLOSE

Reviewer: The server can't guarantee this. (see proposed text below)

Author: A server can't guarantee the packet arrives, but I think sending both guarantees that if it does arrive at the client, the client can be decrypted.  Am I missing something?

a CONNECTION_CLOSE in both Handshake and Initial, because the client discards

> Initial keys as soon as it has Handshake keys. If multiple packets are sent,

Reviewer 4: ```suggestion
Initial keys as soon as it has Handshake keys.  If multiple packets are sent,
```

Author: I believe we've been using one space in this doc?

> (see {{packet-coalesce}}).

Reviewer: Suggested rewording of text after the first sentence:
"Under these circumstances, an endpoint SHOULD send a CONNECTION_CLOSE frame in both Handshake and Initial packets to increase the likelihood of a CONNECTION_CLOSE frame being processed by the peer.  These packets can be coalesced into a single UDP datagram (see {{packet-coalesce}})."

Reviewer 3: A bigger point here is that the server also isn't sure when the client gets 1-RTT keys.  We should say the same about that transition.  I believe that we can rely on handshake confirmation for this transition.  Endpoints might choose to send CONNECTION_CLOSE in Handshake packets prior to the handshake being confirmed.

Author: Agreed, another spot we're relying on #2673?

Author: I updated the text about sending 1-RTT connection closes to use Handshake confirmed, which lines up with #2673.  I think that's a good fix?

Reviewer 3: Don't sound so uncertain :)  I think that the fix is good.

And we'd (collectively) be crazy if we didn't take #2673.

Reviewer 5: We'd also individually go crazy if we don't take it.  😉

Reviewer: "Under these circumstances, a server SHOULD send a CONNECTION_CLOSE frame in both Handshake and Initial packets to ensure that at least one of them is processable by the client. These packets can be coalesced into a single UDP datagram (see {{packet-coalesce}})."

> they can be coalesced (see {{packet-coalesce}}) to facilitate retransmission.

Reviewer 4: ```suggestion
they can be coalesced (see {{packet-coalesce}}).
```

Author: Well, it both minimizes the number of datagrams and makes retransmission easier.  I'll update it with more verbose text about why this can facilitate retransmission.
